src/install/dracut-install.c

```diff
@@ -1515,7 +1515,105 @@ static bool check_module_path(const char *path)
+{
+        DIR *d;
+        struct dirent *dir;
+        d = opendir(node_path);
+        if (d) {
+                while ((dir = readdir(d)) != NULL) {
+                        if (strstr(dir->d_name, "supplier") != NULL) {
+                                char supplier_path[PATH_MAX] = {0};
+                                if (realpath(dir->d_name, supplier_path)) {
+                                        strncat(supplier_path, "/supplier", sizeof(supplier_path));
+
+                                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                                        struct kmod_list *l = NULL;
+
+                                        if (find_module_from_sysfs_node(ctx, supplier_path, &list) >= 0) {
+                                                kmod_list_foreach(l, list) {
+                                                        struct kmod_module *mod = kmod_module_get_module(l);
+                                                        char *supplier_module = strdup(kmod_module_get_name(mod));
+                                                        kmod_module_unref(mod);
+                                                        hashmap_put(suppliers, supplier_module, supplier_module);
+                                                }
+                                        }
+                                }
+                        }
+                }
+                closedir(d);
+        }
+}
+
+static void find_suppliers_for_module(struct kmod_ctx *ctx, Hashmap *suppliers, const char *module)
+{
+        _cleanup_fts_close_ FTS *fts;
+        char *paths[] = { "/sys/devices/platform", NULL };
+        fts = fts_open(paths, FTS_NOSTAT | FTS_PHYSICAL, NULL);
+
+        for (FTSENT *ftsent = fts_read(fts); ftsent != NULL; ftsent = fts_read(fts)) {
+                if (strcmp(ftsent->fts_name, "modalias") == 0) {
+                        char sysfs_node[PATH_MAX] = {0};
+                        memcpy(sysfs_node, ftsent->fts_parent->fts_path, ftsent->fts_parent->fts_pathlen);
+                        sysfs_node[ftsent->fts_parent->fts_pathlen] = '\0';
+
+                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                        struct kmod_list *l = NULL;
+
+                        if (find_module_from_sysfs_node(ctx, sysfs_node, &list) < 0)
+                                continue;
+
+                        kmod_list_foreach(l, list) {
+                                _cleanup_kmod_module_unref_ struct kmod_module *mod = kmod_module_get_module(l);
+                                const char *name = kmod_module_get_name(mod);
+
+                                if (strcmp(name, module) == 0) {
+                                        find_suppliers_for_sys_node(ctx, suppliers, sysfs_node);
+
+                                        // Also add suppliers of parent nodes
+                                        FTSENT *parent = ftsent->fts_parent;
```

Comment:
on my amd64 machine the outer-most loop yields:
```
/sys/devices/platform/PNP0C14:00/wmi_bus/wmi_bus-PNP0C14:00/0E7AF9F2-44A1-4C6F-A4B0-A7678480DA61
/sys/devices/platform/PNP0C14:00
/sys/devices/platform/reg-dummy
/sys/devices/platform/coretemp.1
/sys/devices/platform/dcdbas
/sys/devices/platform/platform-framebuffer.0
/sys/devices/platform/microcode
/sys/devices/platform/GHES.32994
/sys/devices/platform/dmi-ipmi-si.0
/sys/devices/platform/GHES.32992
/sys/devices/platform/GHES.49377
/sys/devices/platform/coretemp.0
/sys/devices/platform/ipmi_bmc.0
/sys/devices/platform/PNP0103:00
/sys/devices/platform/serial8250
/sys/devices/platform/i8042/serio0
/sys/devices/platform/i8042/serio1
/sys/devices/platform/i8042
/sys/devices/platform/GHES.32993
/sys/devices/platform/GHES.49378
/sys/devices/platform/GHES.227
/sys/devices/platform/pcspkr/input/input5
/sys/devices/platform/pcspkr
/sys/devices/platform/GHES.49376
```

Are you *sure* it's safe/sensible to process, for example, `/s/d/p/i8042` thrice? If not, mark `parent->fts_number = 1` on each iteration and make the loop condition `parent->level >= 1 && !parent->fts_number`.